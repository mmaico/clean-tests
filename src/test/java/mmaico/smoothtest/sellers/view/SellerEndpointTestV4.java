package mmaico.smoothtest.sellers.view;


import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import mmaico.smoothtest.infratest.BaseTest;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.test.web.servlet.MvcResult;

import java.io.IOException;

import static java.util.Arrays.asList;
import static mmaico.smoothtest.infratest.BDD.*;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.springframework.http.HttpStatus.CREATED;
import static org.springframework.http.HttpStatus.OK;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;

/**
 * Test with Maturity level 4
 *  Statement BDD OK
 *  Json Inside scenario file OK
 *  Wiremock inside scenario file OK
 *  Match Snapshot OK <---
 *  Custom comparator for dynamic fields NOK
 */
@ExtendWith({SnapshotExtension.class})
class SellerEndpointTestV4 extends BaseTest {


    private Expect expect;

    @BeforeAll
    public static void beforeAll() throws IOException {
        scenarios.load("seller-scenario.json");
    }

    @BeforeEach
    public void setUp() {
        this.expect = expect.serializer("json");
    }

    @SnapshotName("should return seller by id")
    @Test
    public void shouldReturnSellerById() throws Exception {
        Given("a seller with valid id");
            stub("seller with id 53ac008b-9d16-4c36-afea-0e49c0c3515a");
        And("has a velid level");
            stub("seller level with id 53ac008b-9d16-4c36-afea-0e49c0c3515a");
        When("the api is called passing a id");
            MvcResult response = mockMvc.perform(get("/sellers/53ac008b-9d16-4c36-afea-0e49c0c3515a")).andReturn();
        Then("must return the seller with all the data");
            assertEquals(OK.value(), response.getResponse().getStatus());
            expect.toMatchSnapshot(response.getResponse().getContentAsString());
    }

    @SnapshotName("should return seller by id using retry")
    @Test
    public void shouldReturnASellerUsingRetry() throws Exception {
        Given("a seller with valid id");
            stub("Seller search result (start)");
            asList(2, 3, 4).forEach(count -> stub("Seller search result not finished yet (intermediate)", "count", count));
            stub("Ended seller search result (end)");
        And("has a velid level");
            stub("seller level with id 68dcf440-d1fe-4ff0-ae71-9db139f77a8e");
        When("the api is called passing a id");
            MvcResult response = mockMvc.perform(get("/sellers/68dcf440-d1fe-4ff0-ae71-9db139f77a8e")).andReturn();
        Then("must return the seller with all the data");
            expect.toMatchSnapshot(response.getResponse().getContentAsString());
            assertEquals(OK.value(), response.getResponse().getStatus());

    }

    @SnapshotName("should find all sellers registered")
    @Test
    public void shouldFindAllSellersRegistered() throws Exception {
        Given("a sellers api");
            stub("find all sellers registered");
        And("have a seller level default for each request");
            stub("seller level for a list");
        When("the api is called");
            MvcResult response = mockMvc.perform(get("/sellers")).andReturn();
        Then("must return all sellers registered");
            assertEquals(OK.value(), response.getResponse().getStatus());
            expect.toMatchSnapshot(response.getResponse().getContentAsString());
    }

    /**
     * This test will fail because the return a dynamic id, the solution to this problem
     * is in the custom comparator implemented in the main test
     * @throws Exception
     */
    @SnapshotName("should create a new seller")
    @Test
    public void shouldCreateANewSeller() throws Exception {
        Given("a seller with all valid data");
            stub("Create a new seller with all valid parameters");
        When("the api is called passing a id");
            MvcResult response = mockMvc.perform(post("/sellers")
                .contentType("application/hal+json")
                .content(s("seller to be created"))).andReturn();
        Then("must return the seller with all the data");
            assertEquals(CREATED.value(), response.getResponse().getStatus());
            expect.toMatchSnapshot(response.getResponse().getContentAsString());
    }

}