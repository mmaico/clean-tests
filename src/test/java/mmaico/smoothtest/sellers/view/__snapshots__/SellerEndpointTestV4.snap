should create a new seller=[
  "{\"id\":\"290b2ec9-fbe8-4bb1-9160-cadcf9182a23\",\"name\":\"Mandalorian\",\"enrollment\":\"2018-12-06T17:58:16.000+0000\",\"level\":5,\"_links\":{\"self\":{\"href\":\"/sellers/290b2ec9-fbe8-4bb1-9160-cadcf9182a23\"},\"hasContacts\":{\"href\":\"/sellers/290b2ec9-fbe8-4bb1-9160-cadcf9182a23/contacts\"},\"hasScore\":{\"href\":\"/sellers/scores/5524\"}}}"
]


should find all sellers registered=[
  "{\"_embedded\":{\"sellers\":[{\"id\":\"ba262eb0-5178-4a94-9771-49dd77b1c846\",\"name\":\"Baby Yoda\",\"enrollment\":\"2018-12-06T17:58:16.000+0000\",\"level\":4,\"_links\":{\"self\":{\"href\":\"/sellers/ba262eb0-5178-4a94-9771-49dd77b1c846\"},\"hasContacts\":{\"href\":\"/sellers/ba262eb0-5178-4a94-9771-49dd77b1c846/contacts\"},\"hasScore\":{\"href\":\"/sellers/scores/55236\"}}},{\"id\":\"f6724b21-6c23-40a6-960e-233574fad5de\",\"name\":\"Jabba\",\"enrollment\":\"2018-12-06T17:58:16.000+0000\",\"level\":4,\"_links\":{\"self\":{\"href\":\"/sellers/f6724b21-6c23-40a6-960e-233574fad5de\"},\"hasContacts\":{\"href\":\"/sellers/f6724b21-6c23-40a6-960e-233574fad5de/contacts\"},\"hasScore\":{\"href\":\"/sellers/scores/6464564\"}}},{\"id\":\"cd694802-dd85-4248-b80f-d974a2d8dbe4\",\"name\":\"Jango Fett\",\"enrollment\":\"2018-12-06T17:58:16.000+0000\",\"level\":4,\"_links\":{\"self\":{\"href\":\"/sellers/cd694802-dd85-4248-b80f-d974a2d8dbe4\"},\"hasContacts\":{\"href\":\"/sellers/cd694802-dd85-4248-b80f-d974a2d8dbe4/contacts\"},\"hasScore\":{\"href\":\"/sellers/scores/9998887\"}}}]}}"
]


should return seller by id using retry=[
  "{\"id\":\"68dcf440-d1fe-4ff0-ae71-9db139f77a8e\",\"name\":\"Chewbacca\",\"enrollment\":\"2022-12-06T17:58:16.000+0000\",\"level\":9,\"_links\":{\"self\":{\"href\":\"/sellers/68dcf440-d1fe-4ff0-ae71-9db139f77a8e\"},\"hasContacts\":{\"href\":\"/sellers/68dcf440-d1fe-4ff0-ae71-9db139f77a8e/contacts\"},\"hasScore\":{\"href\":\"/sellers/scores/9995562325\"}}}"
]


should return seller by id=[
  "{\"id\":\"53ac008b-9d16-4c36-afea-0e49c0c3515a\",\"name\":\"Boba Fett\",\"enrollment\":\"2018-12-06T17:58:16.000+0000\",\"level\":5,\"_links\":{\"self\":{\"href\":\"/sellers/53ac008b-9d16-4c36-afea-0e49c0c3515a\"},\"hasContacts\":{\"href\":\"/sellers/53ac008b-9d16-4c36-afea-0e49c0c3515a/contacts\"},\"hasScore\":{\"href\":\"/sellers/scores/552544458\"}}}"
]